cli/upgrade/upgrade.go

```diff
@@ -0,0 +1,183 @@
+package upgrade
+
+import (
+	_ "embed"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/tarantool/tt/cli/connector"
+	"github.com/tarantool/tt/cli/replicaset"
+	"github.com/tarantool/tt/cli/running"
+)
+
+//go:embed lua/upgrade.lua
+var upgradeLuaScript string
+
+type SyncInfo struct {
+	LSN uint32  `mapstructure:"lsn"`
+	IID uint32  `mapstructure:"iid"`
+	Err *string `mapstructure:"err"`
+}
+
+func internalUpgrade(conn connector.Connector) (SyncInfo, error) {
+	var upgradeInfo SyncInfo
+	res, err := conn.Eval(upgradeLuaScript, []any{}, connector.RequestOpts{})
+	if err != nil {
+		return upgradeInfo, err
+	}
+	if err := mapstructure.Decode(res[0], &upgradeInfo); err != nil {
+		return upgradeInfo, err
+	}
+	if upgradeInfo.Err != nil {
+		return upgradeInfo, errors.New(*upgradeInfo.Err)
+	}
+	return upgradeInfo, nil
+}
+
+func WaitLSN(conn connector.Connector, masterUpgradeInfo SyncInfo, timeout int) error {
+	for i := 0; i < timeout; i++ {
+		var res []interface{}
+		res, err := conn.Eval(
+			fmt.Sprintf("return box.info.vclock[%d]", masterUpgradeInfo.IID),
+			[]any{}, connector.RequestOpts{})
+		if err != nil || len(res) == 0 {
+			return err
+		}
```

Comment:
It is unclear, whether the `err == nil && len(res) == 0` case is supposed to be a successful result or an error. It looks like this `if` returns an error (it has `return err`), but `err` is `nil` in the case.

Personally, I often use the following pattern (pseudocode):

```go
var res, err = foo()
// First failure type.
if <predicate 1> {
    return err
}
// Second failure type.
if <predicate 2> {
    return err
}
// Success, exit earlier.
if <predicate 3> {
    return res
}
// Continue retrying...
```

The idea is to visually split handling of different `foo`'s exit situations (even if the actions to handle them are the same). It takes more lines, but easier to read.

Reply:
Fixed.